controllers/etcdpeer_controller.go

```diff
@@ -22,13 +24,22 @@ type EtcdPeerReconciler struct {
 }
 
 const (
-	etcdImage = "quay.io/coreos/etcd:v3.2.27"
+	etcdImage                = "quay.io/coreos/etcd:v3.2.27"
+	etcdInitialClusterEnvVar = "ETCD_INITIAL_CLUSTER"
 )
 
 // +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=apps,resources=replicaset,verbs=get;update;patch;create
 
+func clusterDiscoveryString(bootstrapPeers []etcdv1alpha1.BootstrapPeer) string {
+	s := make([]string, len(bootstrapPeers))
+	for i, bp := range bootstrapPeers {
+		s[i] = fmt.Sprintf("%s=http://%s:2380", bp.Name, bp.Host)
+	}
+	return strings.Join(s, ",")
+}
```

Comment:
This logic shouldn't be in the EtcdPeer controller, it should live in the (eventual) EtcdCluster controller. The EtcdPeer resource type should be 'dumb', as otherwise it'll be harder for us to implement things like backup/restore, and joining existing clusters not managed by the operator (if we ever support this).

The less magic the better 😄

Reply:
Shouldn't it? Does that mean that the `EtcdPeer` resource should take only the final bootstrap string as a simple `string`, and not an array of `bootstrapPeers` at all? That's all this function does - join together the `bootstrapPeer` entries into a string for passing down as an envvar.

If we did this on the cluster, that means that the cluster operator is now very knowledgable about how the peers are configured and the format of etcd's environment variables.

Reply:
Sorry, I thought I'd responded to this...

I was referring to the hardcoded `http` and `2380`, not the fact it flattens a struct into a comma separated list. I agree that flattening the []struct to a single `string` should be the responsibility of the `etcdpeer` controller 😄

Reply:
I think this can be resolved in favour of the other thread on this issue